web/src/main/java/com/marklogic/hub/web/service/FlowJobService.java

```diff
@@ -56,127 +55,132 @@ public FlowJobService() {
-        if (this.client == null) {
-            this.setupClient();
+    public FlowJobs getJobs(String flowName, boolean forceRefresh) {
+        if (forceRefresh) {
+            FlowJobs flowJobs = retrieveJobsByFlowName(flowName);
+            cachedJobsByFlowName.put(flowName, flowJobs);
+            return flowJobs;
         }
         try {
-            return cachedJobsByFlowName.get(flowName, () -> {
-                client.init(ML_JOBS_NAME, this);
-                RequestParameters params = new RequestParameters();
-                if (StringUtils.isNotEmpty(flowName)) {
-                    params.add("flow-name", flowName);
-                }
+            return cachedJobsByFlowName.get(flowName, () -> retrieveJobsByFlowName(flowName));
+        } catch (Exception e) {
+            logger.error(e.getMessage());
+        } finally {
+            if (cachedJobsByFlowName.getIfPresent(flowName) == null) {
+                cachedJobsByFlowName.invalidate(flowName);
+            }
+        }
 
-                ServiceResultIterator resultItr = this.getServices().get(params);
-                if (resultItr == null || !resultItr.hasNext()) {
-                    throw new RuntimeException("Unable to get job document");
-                }
-                ServiceResult res = resultItr.next();
-                JsonNode jsonNode = res.getContent(new JacksonHandle()).get();
-
-                Flow flow = flowManager.getFlow(flowName);
-                Map<String, Step> steps = flow.getSteps();
-
-                List<String> jobs = new ArrayList<>();
-                LatestJob latestJob = new LatestJob();
-                FlowJobs flowJobs = new FlowJobs(jobs, latestJob);
-
-                final JsonNode[] lastJob = {null};
-                final String[] lastTime = {null, null}; //store last start and end time
-                if (jsonNode.isArray()) {
-                    jsonNode.forEach(job -> {
-                        JSONObject jobJson = new JSONObject(job.get("job"));
-                        String jobId = jobJson.getString("jobId");
-                        jobs.add(jobId);
-                        String currStartTime = jobJson.getString("timeStarted", "");
-                        String currEndTime = jobJson.getString("timeEnded", "");
-                        if (StringUtils.isEmpty(lastTime[0])) {
+        return cachedJobsByFlowName.getIfPresent(flowName);
+    }
+
+    private FlowJobs retrieveJobsByFlowName(String flowName) {
+        if (client == null) {
+            setupClient();
+        }
+        client.init(ML_JOBS_NAME, this);
+        RequestParameters params = new RequestParameters();
+        if (StringUtils.isNotEmpty(flowName)) {
+            params.add("flow-name", flowName);
+        }
+
+        ServiceResultIterator resultItr = this.getServices().get(params);
+        if (resultItr == null || !resultItr.hasNext()) {
+            throw new RuntimeException("Unable to get job document");
```

Comment:
I recommend tossing the flow name into this message (would it really ever be null, given the method name?) - e.g. "No jobs found for flow with name: " + flowName